File: src/java/leetcode/editor/en/P987_VerticalOrderTraversalOfABinaryTree.java
package java.leetcode.editor.en;
import java.util.*;
import java.util.function.*;
import java.leetcode.editor.en.definition.*;

public class P987_VerticalOrderTraversalOfABinaryTree {
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private Map<Integer, Map<Integer, PriorityQueue<Integer>>> map = new TreeMap<>();
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        dfs(root, 0, 0);
        List<List<Integer>> ans = new ArrayList<>();
        for (Map<Integer, PriorityQueue<Integer>> m : map.values()) {
            List<Integer> list = new ArrayList<>();
            for (PriorityQueue<Integer> pq : m.values()) {
                while (!pq.isEmpty()) list.add(pq.poll());
            }
            ans.add(list);
        }
        return ans;
    }
    private void dfs(TreeNode root, int x, int y) {
        if (root == null) return;
        map.putIfAbsent(x, new TreeMap<>());
        map.get(x).putIfAbsent(y, new PriorityQueue<>());
        map.get(x).get(y).offer(root.val);
        dfs(root.left, x-1, y+1);
        dfs(root.right, x+1, y+1);
    }
}
//leetcode submit region end(Prohibit modification and deletion)
}

File: src/kt/leetcode/editor/en/P987_VerticalOrderTraversalOfABinaryTree.kt
package kt.leetcode.editor.en;
import java.util.*;
import java.util.function.*;
import kt.leetcode.editor.en.definition.*;

public class P987_VerticalOrderTraversalOfABinaryTree {
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun verticalTraversal(root: TreeNode?): List<List<Int>> {
        val ans = ArrayList<ArrayList<Int>>()
        val map = TreeMap<Int, TreeMap<Int, PriorityQueue<Int>>>()
        /*  nested function begin */
        fun dfs(root: TreeNode?, x: Int, y: Int) {
            if (root == null) return
            map.putIfAbsent(x, TreeMap())
            map[x]!!.putIfAbsent(y, PriorityQueue())
            map[x]!![y]!!.add(root.`val`)
            dfs(root.left, x-1, y+1)
            dfs(root.right, x+1, y+1)
        }
        /*  nested function end */
        dfs(root, 0, 0)
        map.values.forEach { mapByX ->
            ans.add(ArrayList())
            mapByX.values.forEach { pq ->
                while (pq.isNotEmpty()) ans.last().add(pq.poll())
            }
        }
        return ans
    }
}
//leetcode submit region end(Prohibit modification and deletion)
}
